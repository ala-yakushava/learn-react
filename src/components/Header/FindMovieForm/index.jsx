import React from 'react';

import './style.scss';
import Heading from '../../Heading';
import TextInput from '../../TextInput';
import Button from '../../Button';

class FindMovieForm extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      text: '',
    };
  }

  handleInputChange = () => (evt) => {
    const { target: { name, value } } = evt;
    this.setState({
      [name]: value,
    });
  }

  handleSubmit = (evt) => {
    console.log(this.state);
    evt.preventDefault();
  }

  render() {
    return (
      <form className="FindMovieForm" onSubmit={this.handleSubmit}>
        <Heading className="FindMovieForm_title">Find your movie</Heading>
        <TextInput
          name="text"
          className="FindMovieForm_input"
          placeholder="What do you want to watch?"
          value={this.state.text}
          isRequired={true}
          onChange={this.handleInputChange(this.state.text)}
        />
        <Button type="submit" mode="primary">Search</Button>
      </form>
    );
  }
}

export default FindMovieForm;
