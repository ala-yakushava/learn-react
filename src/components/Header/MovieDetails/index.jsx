import React from 'react';
import PropTypes from 'prop-types';

import './style.scss';
import Heading from '../../Heading';

const MovieDetails = ({ movie }) => {
  const {
    title, tagline, overview, runtime, vote_average: vote, release_date: date, poster_path: poster,
  } = movie;

  const releaseYear = date ? date.split('-')[0] : null;

  return (
    <article className="MovieDetails" data-testid="movie-details">
      <div className="MovieDetails_picture">
        <img src={poster} alt={`Poster for the movie ${title}`} />
      </div>
      <div className="MovieDetails_inner">
        <header className="MovieDetails_header">
          <Heading className="MovieDetails_title">{ title }</Heading>
          <span className="MovieDetails_vote">{ vote }</span>
        </header>
        <p className="MovieDetails_lead">{ tagline }</p>
        <p className="MovieDetails_details">
          <span>{ releaseYear }</span>
          <span>{ `${runtime} min` }</span>
        </p>
        <p className="MovieDetails_description">{ overview }</p>
      </div>
    </article>
  );
};

MovieDetails.propTypes = {
  movie: PropTypes.shape({
    title: PropTypes.string,
    tagline: PropTypes.string,
    vote_average: PropTypes.number,
    runtime: PropTypes.number,
    release_date: PropTypes.string,
    poster_path: PropTypes.string,
    overview: PropTypes.string,
  }),
};

MovieDetails.defaultProps = {
  movie: {},
};

export default MovieDetails;
