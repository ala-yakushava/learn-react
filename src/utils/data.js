import uniqueId from 'lodash/uniqueId';

export const sortItems = [
  {
    id: 'sort-rating',
    value: 'vote_average',
    label: 'Rating',
  },
  {
    id: 'sort-release_date',
    value: 'release_date',
    label: 'Release date',
  },
];

export const filterItems = [
  {
    id: 'filter-all',
    value: 'All',
    label: 'All',
  },
  {
    id: 'filter-documentary',
    value: 'Documentary',
    label: 'Documentary',
  },
  {
    id: 'filter-comedy',
    value: 'Comedy',
    label: 'Comedy',
  },
  {
    id: 'filter-horror',
    value: 'Horror',
    label: 'Horror',
  },
  {
    id: 'filter-crime',
    value: 'Crime',
    label: 'Crime',
  },
];

export const formField = [
  {
    id: uniqueId(),
    name: 'title',
    type: 'text',
    label: 'Title',
    placeholder: 'Title here',
  },
  {
    id: uniqueId(),
    name: 'release_date',
    type: 'date',
    label: 'Release date',
    placeholder: 'Select date',
  },
  {
    id: uniqueId(),
    name: 'poster_path',
    type: 'url',
    label: 'Movie URL',
    placeholder: 'Movie URL here',
  },
  {
    id: uniqueId(),
    name: 'genres',
    type: 'text',
    label: 'Genre',
    placeholder: 'Select genre',
  },
  {
    id: uniqueId(),
    name: 'overview',
    type: 'text',
    label: 'Overview',
    placeholder: 'Overview text goes here',
  },
  {
    id: uniqueId(),
    name: 'runtime',
    type: 'text',
    label: 'Runtime',
    placeholder: 'Runtime text goes here',
  },
];
