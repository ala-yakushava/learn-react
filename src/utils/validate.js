import * as yup from 'yup';

const MovieFormSchema = yup.object().shape({
  title: yup.string().transform((value) => value.trim()).min(2).required('Required'),
  release_date: yup.date().max(new Date(), 'Future date'),
  poster_path: yup.string().url('Invalid URL').required('Required'),
  genres: yup.string().transform((value) => value.trim()).min(3).required('Required'),
  overview: yup.string().transform((value) => value.trim()).min(10).required('Required'),
  runtime: yup.number('Invalid number').required('Required').min(0, 'Too Short!').integer(),
});

export default MovieFormSchema;
