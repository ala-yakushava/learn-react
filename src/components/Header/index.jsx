import React from 'react';

import './style.scss';
import Logo from '../Logo';
import Button from '../Button';
import FindMovieForm from './FindMovieForm';
import ModalPortal from '../ModalPortal';
import Modal from '../Modal';
import AddMovieForm from '../Modal/AddMovieForm';

class Header extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      visible: false,
    };
  }

  handleClick = () => {
    this.setState(state => ({
      visible: true,
    }));
  }

  handleClickCloseButton = () => {
    this.setState(state => ({
      visible: false,
    }));
  }

  render() {
    return (
      <header className="Header">
        <Logo />
        <Button mode="transparent" onClick={this.handleClick}>Add movie</Button>
        <FindMovieForm className="Header_form"/>
        {this.state.visible &&
          <ModalPortal>
            <Modal onClick={this.handleClickCloseButton}>
              <AddMovieForm />
            </Modal>
          </ModalPortal>
        }
      </header>
    );
  }
};

export default Header;
