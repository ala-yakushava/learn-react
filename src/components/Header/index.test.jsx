import React from 'react';
import { useSelector, useDispatch } from 'react-redux';
import { MemoryRouter } from 'react-router-dom';
import renderer from 'react-test-renderer';

import Header from './index';
import { id, movie } from '../../utils/mock-data';

jest.mock('react-redux');

describe('Components - Header', () => {
  const dispatch = jest.fn();

  beforeAll(() => {
    useSelector.mockImplementation((cb) => cb({ movieInfo: { movie } }));
    useDispatch.mockReturnValue(dispatch);
  });

  afterAll(() => {
    useSelector.mockClear();
    useDispatch.mockClear();
  });

  test('should match snapshot', () => {
    const tree = renderer
      .create(
        <MemoryRouter initialEntries={[`/movies/${id}`]}>
          <Header />
        </MemoryRouter>,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
