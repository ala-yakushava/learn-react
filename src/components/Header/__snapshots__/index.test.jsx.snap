// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Components - Header should match snapshot 1`] = `
<header
  className="Header"
>
  <p
    className="Logo"
  >
    <b>
      netflix
    </b>
    roulette
  </p>
  <button
    className="Button Button--transparent"
    disabled={false}
    onClick={[Function]}
    type="button"
  >
    <span
      className="Button_text"
    >
      Search
    </span>
  </button>
  <div
    className="Header_inner"
  >
    <article
      className="MovieDetails"
      data-testid="movie-details"
    >
      <div
        className="MovieDetails_picture"
      >
        <img
          alt="Poster for the movie Thor: Ragnarok"
          src="https://image.tmdb.org/t/p/w500/rzRwTcFvttcN1ZpX2xv4j3tSdJu.jpg"
        />
      </div>
      <div
        className="MovieDetails_inner"
      >
        <header
          className="MovieDetails_header"
        >
          <div
            className="Heading MovieDetails_title"
          >
            Thor: Ragnarok
          </div>
          <span
            className="MovieDetails_vote"
          >
            7.4
          </span>
        </header>
        <p
          className="MovieDetails_lead"
        >
          No Hammer. No Problem.
        </p>
        <p
          className="MovieDetails_details"
        >
          <span>
            2017
          </span>
          <span>
            130 min
          </span>
        </p>
        <p
          className="MovieDetails_description"
        >
          Thor is imprisoned on the other side of the universe and finds himself in a race against time to get back to Asgard to stop Ragnarok, the prophecy of destruction to his homeworld and the end of Asgardian civilization, at the hands of an all-powerful new threat, the ruthless Hela.
        </p>
      </div>
    </article>
  </div>
</header>
`;
